### ISSUE TITLE
Using spdk_pci_device_detach () detach The pcie device is blocked

### ISSUE THREAD

**Issue by xiankelai00001** [spdk/spdk#2652]
# Bug report

## Possible Solution

gdb info

Thread 1 "debug_pci_device_detach" hit Breakpoint 1, spdk_pci_device_detach (dev=0xaf3730) at pci.c:526
526             assert(dev->internal.attached);
Missing separate debuginfos, use: yum debuginfo-install libuuid-2.32.1-28.el8.x86_64 numactl-libs-2.0.12-13.el8.x86_64 openssl-libs-1.1.1k-5.el8_5.x86_64 zlib-1.2.11-17.el8.x86_64
(gdb) n
528             if (dev->internal.claim_fd >= 0) {
(gdb) p dev->internal.attached
$1 = true
(gdb) s
532             dev->internal.attached = false;
(gdb) s
533             if (strcmp(dev->type, "pci") == 0) {
(gdb) s
541                     detach_rte(dev);
(gdb) s
detach_rte (dev=0xaf3730) at pci.c:125
125             struct rte_pci_device *rte_dev = dev->dev_handle;
(gdb) s
129             if (!spdk_process_is_primary()) {
(gdb) s
spdk_process_is_primary () at env.c:295
295             return (rte_eal_process_type() == RTE_PROC_PRIMARY);
(gdb) n
296     }
(gdb) n
detach_rte (dev=0xaf3730) at pci.c:129
129             if (!spdk_process_is_primary()) {
(gdb) s
134             pthread_mutex_lock(&g_pci_mutex);
(gdb) s

**I guess pthread_mutex_lock(&g_pci_mutex) is stuck**

## Steps to Reproduce

1.   spdk_pci_enumerate(spdk_pci_nvme_get_driver(), enum_cb, NULL);
2.  spdk_pci_for_each_device(NULL, pcie_detach_cb)
3.  Calling spdk_pci_device_detach() in the pcie_detach_cb function gets stuck
4.

## Context (Environment including OS version, SPDK version, etc.)
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
Linux localhost 4.18.0-305.25.1.el8_4.x86_64 #1 SMP Wed Nov 3 10:29:07 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

[2022-08-19 14:19:40.201593768] NOTIC| Starting SPDK v22.09-pre git sha1 50e8ac207 / DPDK 22.03.0 initialization...
[2022-08-19 14:19:40.201608341] NOTIC| [ DPDK EAL parameters: 
[2022-08-19 14:19:40.201611185] NOTIC| nvme 
[2022-08-19 14:19:40.201613038] NOTIC| -c 0xfffff 
[2022-08-19 14:19:40.201614787] NOTIC| -m 1024 
[2022-08-19 14:19:40.201617720] NOTIC| --log-level=lib.eal:6 
[2022-08-19 14:19:40.201619358] NOTIC| --log-level=lib.cryptodev:5 
[2022-08-19 14:19:40.201620988] NOTIC| --log-level=user1:6 
[2022-08-19 14:19:40.201622582] NOTIC| --iova-mode=pa 
[2022-08-19 14:19:40.201624225] NOTIC| --base-virtaddr=0x200000000000 
[2022-08-19 14:19:40.201625846] NOTIC| --match-allocations 
[2022-08-19 14:19:40.201640842] NOTIC| --file-prefix=spdk76379 
[2022-08-19 14:19:40.201642460] NOTIC| --proc-type=auto 
[2022-08-19 14:19:40.201644048] NOTIC| ]
TELEMETRY: No legacy callbacks, legacy socket not created

**Comment by Comphix** [comment]
Thanks for reporting this issue.

Just make sure this issue is reproduceable with SPDK nvme tools like identify and so on? Or it is found in your application with SPDK NVMe driver?

**Comment by xiankelai00001** [comment]
My debug Code:
![image](https://user-images.githubusercontent.com/31873486/185568147-c239cecb-a352-4e82-a378-5ee6d4ccfb23.png)

**Comment by Comphix** [comment]
You may also refer to ./app/spdk_lspci/spdk_lspci.c. Also may take a look at the ./examples/nvme/identify/identify.c

**Comment by xiankelai00001** [comment]
I have read it (./app/spdk_lspci/spdk_lspci.c)

Here's a question:
Using the "spdk_pci_enumerate " function to enumerate pcie devices, does he not need to detach?
"Attached devices have to be manually detached with spdk_pci_device_detach() to be attach-able Again. "This description comes from spdk_pci_enumerate.

My Needs:
When cc.en=0, operate the pcie register. I do not want cc.en=1 after operating the pcie register.

My thoughts:
Step 1: Use the spdk_pCI_enumerate function to enumerate pcie devices.
Step 2: Use spdk_pCI_for_each_device and spdk_pCI_device_detach to detach pcie devices
Step 3: Use spdk_nvme_probe to initiate the controller.

Results:
It got stuck in the second step.

**Comment by changpe1** [comment]
I don't follow your requirement and issue.  You need to give a detailed description and reproduce steps.

**Comment by xiankelai00001** [comment]
reproduce steps:

![image](https://user-images.githubusercontent.com/31873486/185583311-b388c3cb-4ea0-460c-b996-907879fde66d.png)
![image](https://user-images.githubusercontent.com/31873486/185583471-a2fc5621-bd3a-42dd-b98f-733066517331.png)

**Comment by jimharris** [comment]
I don't see any reason why the g_pci_mutex could not be made recursive.  That would allow spdk_pci_device_detach() to be called from an spdk_pci_for_each_device callback.

**Comment by xiankelai00001** [comment]
I think the fundamental reason for this problem: "pthread_mutex_lock (& g_pci_mutex);" The mutual lock was called twice.

the code comes from "https://review.spdk.io/gerrit/c/spdk/spdk/+/13956"
`void
spdk_pci_for_each_device(void *ctx, void (*fn)(void *ctx, struct spdk_pci_device *dev))
{
	struct spdk_pci_device *dev, *tmp;

	**pthread_mutex_lock(&g_pci_mutex);**
	TAILQ_FOREACH_SAFE(dev, &g_pci_devices, internal.tailq, tmp) {
		fn(ctx, dev);
	}
	pthread_mutex_unlock(&g_pci_mutex);
}`

"g_pci_mutex" was locked before "TAILQ_FOREACH_SAFE".
In the ”spdk_pci_device_detach“ function, call "pthread_mutex_lock (& g_pci_mutex)" to be starved to death。

`void
spdk_pci_device_detach(struct spdk_pci_device *dev)
{
	assert(dev->internal.attached);

	if (dev->internal.claim_fd >= 0) {
		spdk_pci_device_unclaim(dev);
	}

	dev->internal.attached = false;
	if (strcmp(dev->type, "pci") == 0) {
		/* if it's a physical device we need to deal with DPDK on
		 * a different process and we can't just unset one flag
		 * here. We also want to stop using any device resources
		 * so that the device isn't "in use" by the userspace driver
		 * once we detach it. This would allow attaching the device
		 * to a different process, or to a kernel driver like nvme.
		 */
		**detach_rte(dev);**                                     **_//// This function uses "pthread_mutex_lock (& g_pci_mutex)"_**
	}

	**cleanup_pci_devices();**                                 **_////  This function uses "pthread_mutex_lock (& g_pci_mutex)"_**
}
`

**Conclusion**: If we want to process a device from within the spdk_pci_for_each_device() callback, then "pthread_mutex_lock(&G_pCI_mutex)" cannot be called again in the callback.
This is a bug caused by improper use of mutual locks(pthread_mutex_lock (&g_pci_mutex)).

**Comment by ksztyber** [comment]
Yeah, I'm not saying that `TAILQ_FOREACH_SAFE` fixes this problem - I agree that the mutex would have to be made recursive. But once it is recursive, you still need to use the _SAFE loop to make sure that you're not touching freed memory if you detach a device from the `spdk_pci_for_each_device()` callback.

**Comment by xiankelai00001** [comment]
I understand what you mean.
I agree with your opinion.

**Comment by ksztyber** [comment]
[Bug scrub] We currently don't support that. We should document that. Marking as an enhancement, because it might be useful to change the current behavior. @xiankelai00001 feel free to pick it up.
